PurityReviewer/AppComponents/utils.py

```diff
@@ -1,450 +1,467 @@
+    mod_tab_df['SCNA_likelihood'] = mode_tab[:, 15]
+    mod_tab_df['Kar_likelihood'] = mode_tab[:, 17]
+    mod_tab_df['SSNVs_likelihood'] = mode_tab[:, 20]
+
+    # Maf file
+    maf_file = rdata_tables.rx2('mut.cn.dat')
+    mut_annot_list = mode_res.rx2('modeled.muts')
+
+    return mod_tab_df,maf_file,mut_annot_list
+
+def calculate_multiplicity(maf,alpha):
+    # Local copy number
+    q = maf['q_hat']
+    # Allele fraction
+    af = maf['alt']/(maf['alt']+maf['ref'])
+
+    # af = (alpha * mult) / (alpha * q + (1-alpha)*2)
+    mult = af * (alpha*q + (1-alpha)*2) / alpha
+
+    return(mult)
+
+def validate_purity(x):
+    return (x >=0) and (x <= 1)
+
+
+def validate_ploidy(x):
+    return (x >=0)
+
+
+def download_data(file_to_download_path, full_local_path):
+    """
+    Downloads data from remote url
+
+    Parameters
+    ==========
+    file_to_download_path: str
+        remote url to file to download
+        
+    full_local_path: str
+        local path to download data
+        
+    """
+    dalmatian.getblob(file_to_download_path).download_to_filename(full_local_path)
+
+
+def download_rdata(rdata_fn_s, rdata_dir, force_download=False):
+    """
+    Download ABSOLUTE rdata locally
+    
+    Parameters
+    ==========
+    rdata_fn_s: pd.Series
+        Pandas series with sample/pair ids in the index and lists their corresponding rdata remote urls
+        
+    rdata_dir: str, Path
+        Local path to directory to save local rdata files
+        
+    force_download: bool
+        To force redownload even if the file already exists locally (by file name)
+
+    Returns
+    =======
+    pd.Series
+        Pandas series with sample/pair ids in the index and lists their corresponding rdata local urls
+    """
+    if not os.path.isdir(rdata_dir):
+        os.mkdir(rdata_dir)
+
+    local_rdata_dict = {}
+    for pair_id, rdata_fn in rdata_fn_s.items():
+        absolute_rdata_gsurl = rdata_fn
+
+        local_absolute_rdata_fn = f'{rdata_dir}/{pair_id}.absolute.rdata'
+        if not os.path.exists(local_absolute_rdata_fn) or force_download:
+            download_data(absolute_rdata_gsurl, local_absolute_rdata_fn)
+
+        local_rdata_dict[pair_id] = local_absolute_rdata_fn
+
+    return pd.Series(local_rdata_dict)
+
+def merge_pairs_precalled_purities_df(pairs_df, precalled_purity_df):
+    """
+    Merges the pairs dataframe with the precalled purity dataframe based on the tumor sample id
+
+    pairs_df: df, dataframe of the pairs sample data
+
+    precalled_purity_df: df, dataframe of the precalled purity values with sample id
+
+    Returns:
+        Merged dataframe with the precalled purity values added as a new column to the original pairs_df dataframe
+    """
+
+    pairs_df = pairs_df.copy()
+    precalled_purity_df = precalled_purity_df.copy()
+
+    return pairs_df.merge(precalled_purity_df, how="inner", left_on="case_sample", right_on="sample_id").set_index('tumor_submitter_id')

```

Comment:
split this up (make the changes in one line and then just return the df)

Reply:
updated to have on two lines